gdal/frmts/rmf/rmflzw.cpp

```diff
@@ -109,23 +111,77 @@ static void LZWUpdateTab(LZWStringTab *poCodeTab, GUInt32 iPred, char bFoll)
     poCodeTab[nNext].bUsed = true;
     poCodeTab[nNext].iNext = 0;
     poCodeTab[nNext].iPredecessor = iPred;
-    poCodeTab[nNext].iFollower = bFoll;
+    poCodeTab[nNext].iFollower = static_cast<GByte>(bFoll);
+}
+
+/************************************************************************/
+/*                            LZWFindIndex()                            */
+/************************************************************************/
+
+static GUInt32 LZWFindIndex(const LZWStringTab* poCodeTab, GUInt32 iPred, char bFoll)
+{
+    GUInt32 nLocal = (iPred + bFoll) | 0x0800;
+    nLocal = (nLocal*nLocal >> 6) & 0x0FFF;      // middle 12 bits of result
+
+    do
+    {
+        CPLAssert(nLocal < TABSIZE);
+        if(poCodeTab[nLocal].iPredecessor == iPred &&
+           poCodeTab[nLocal].iFollower == static_cast<GByte>(bFoll))
+        {
+            return nLocal;
+        }
+        nLocal = poCodeTab[nLocal].iNext;
+    }
+    while(nLocal > 0);
+
+    return NOT_FND;
+}
+
+/************************************************************************/
+/*                             LZWPutCode()                             */
+/************************************************************************/
+
+static bool LZWPutCode(GUInt32 iCode, GUInt32& iTmp, bool& bBitsleft,
+                       GByte*& pabyCurrent, const GByte * const pabyOutEnd)
+{
+    if(bBitsleft)
+    {
+        if(pabyCurrent >= pabyOutEnd)
+        {
+            return false;
+        }
+        *(pabyCurrent++) = static_cast<GByte>((iCode >> 4) & 0xFF);
+        iTmp = iCode & 0x000F;
+        bBitsleft = false;
+    }
+    else
+    {
+        if(pabyCurrent + 1 >= pabyOutEnd)
+        {
+            return false;
+        }
+        *(pabyCurrent++) = static_cast<GByte>(((iTmp << 4) & 0xFF0) +
+                                              ((iCode >> 8) & 0x00F));
+        *(pabyCurrent++) = static_cast<GByte>(iCode & 0xFF);
+        bBitsleft = true;
+    }
+    return true;
 }
 
 /************************************************************************/
 /*                           LZWDecompress()                            */
 /************************************************************************/
 
-int RMFDataset::LZWDecompress( const GByte* pabyIn, GUInt32 nSizeIn,
-                               GByte* pabyOut, GUInt32 nSizeOut,
-                               GUInt32, GUInt32 )
+size_t RMFDataset::LZWDecompress(const GByte* pabyIn, GUInt32 nSizeIn,
+                                 GByte* pabyOut, GUInt32 nSizeOut,
+                                 GUInt32, GUInt32 )
 {
     if( pabyIn == nullptr ||
         pabyOut == nullptr ||
-        nSizeOut < nSizeIn ||
         nSizeIn < 2 )
         return 0;
-
+    GByte* const pabyOutBegin = pabyOut;
     // Allocate space for the new table and pre-fill it
     LZWStringTab *poCodeTab =
         (LZWStringTab *)CPLMalloc( TABSIZE * sizeof(LZWStringTab) );
```

Comment:
the below test is useless

Reply:
@rouault  What is better to use next time CPLMalloc or VSIMalloc? Any suggestions?

Reply:
CPLMalloc() is fine for small allocations (let's say ~ 1 KB) that should normally not fail